src/main/java/net/consensys/orion/http/handler/send/SendHandler.java

```diff
@@ -66,110 +67,109 @@ public SendHandler(
-          "legacy",
-          "Privacy groups to support the creation of groups by privateFor and privateFrom",
-          PrivacyGroupPayload.State.ACTIVE,
-          PrivacyGroupPayload.Type.LEGACY,
-          null);
-
-      final QueryPrivacyGroupPayload queryPrivacyGroupPayload =
-          new QueryPrivacyGroupPayload(keys.toArray(new String[0]), null);
-      final String key = queryPrivacyGroupStorage.generateDigest(queryPrivacyGroupPayload);
-
-
-      queryPrivacyGroupStorage.get(key).thenApply(privacyGroup -> {
-        if (privacyGroup.isPresent()) {
-          send(routingContext, sendRequest, fromKey, toKeys, privacyGroupPayload);
-          return key;
-        } else {
-          return privacyGroupStorage.put(privacyGroupPayload).thenApply((result) -> {
-            queryPrivacyGroupPayload.setPrivacyGroupToAppend(privacyGroupStorage.generateDigest(privacyGroupPayload));
-            return queryPrivacyGroupStorage.update(key, queryPrivacyGroupPayload).thenApply((res) -> {
-              send(routingContext, sendRequest, fromKey, toKeys, privacyGroupPayload);
-              return result;
-            }).exceptionally(e -> {
-              handleFailure(routingContext, e, ErrorType.FIND_GROUP);
-              return null;
-            });
+  private void handleLegacySendRequest(RoutingContext routingContext, SendRequest sendRequest, PublicKey fromKey) {
+    final List<PublicKey> toKeys = Arrays.stream(sendRequest.to()).map(enclave::readKey).collect(Collectors.toList());
+    final ArrayList<String> keys = new ArrayList<>(Arrays.stream(sendRequest.to()).collect(Collectors.toList()));
+    if (sendRequest.from().isPresent()) {
+      keys.add(sendRequest.from().get());
+    }
+    final PrivacyGroupPayload privacyGroupPayload = new PrivacyGroupPayload(
+        keys.toArray(new String[0]),
+        "legacy",
+        "Privacy groups to support the creation of groups by privateFor and privateFrom",
+        PrivacyGroupPayload.State.ACTIVE,
+        PrivacyGroupPayload.Type.LEGACY,
+        null);
+
+    final QueryPrivacyGroupPayload queryPrivacyGroupPayload =
+        new QueryPrivacyGroupPayload(keys.toArray(new String[0]), null);
+    final String key = queryPrivacyGroupStorage.generateDigest(queryPrivacyGroupPayload);
+
+    queryPrivacyGroupStorage.get(key).thenApply(privacyGroup -> {
+      if (privacyGroup.isPresent()) {
+        sendPayloadToParticipants(routingContext, sendRequest, fromKey, toKeys, privacyGroupPayload);
+        return key;
+      } else {
+        return privacyGroupStorage.put(privacyGroupPayload).thenApply((result) -> {
+          queryPrivacyGroupPayload.setPrivacyGroupToAppend(privacyGroupStorage.generateDigest(privacyGroupPayload));
+          return queryPrivacyGroupStorage.update(key, queryPrivacyGroupPayload).thenApply((res) -> {
+            sendPayloadToParticipants(routingContext, sendRequest, fromKey, toKeys, privacyGroupPayload);
+            return result;
```

Comment:
This conditional is probably written backwards - it should deal with "If the group does not exist, create it" - then do the send afterwards. (rather than having the "sendPayloadToParticipants" invoked in 2 different places)